### ISSUE TITLE
prefer-const false positive when assigning specific object keys instead of overwriting with a new value

### ISSUE THREAD

**Issue by gandazgul** [eslint/eslint#7263]
* **ESLint Version:** 2.7.0
* **Node Version:** v6.3.1
* **npm Version:** 3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

```
module.exports = {
    extends: [
        'eslint-config-airbnb',
    ],

    ecmaFeatures: {
        jsx: true,
        modules: true,
    },

    env: {
        browser: true,
        es6: true,
        node: true,
    },

    parser: 'babel-eslint',

    globals: {
        __APP_VERSION__: true,
    },

    rules: {
        'no-undef': 2,
        'id-length': 0,
        'max-len': 0,
        indent: [2, 4, {
            SwitchCase: 1,
        }],

        'brace-style': [
            1,
            'stroustrup',
            {allowSingleLine: true},
        ],

        curly: 2,

        'no-use-before-define': [1, 'nofunc'],
        'no-unused-vars': [2, {args: 'none'}],
        // allowShortCircuit allows for these: callback && callback();
        'no-unused-expressions': [2, {allowShortCircuit: true}],
        'object-curly-spacing': 0,

        'react/jsx-boolean-value': 0,
        'react/jsx-indent-props': [2, 4],
        'react/jsx-no-bind': 0,
        'react/prefer-stateless-function': 0,

        'react/sort-comp': [2, {
            order: [
                'setup',
                'lifecycle',
                'everything-else',
                'render',
            ],
            groups: {
                setup: [
                    'displayName',
                    'propTypes',
                    'contextTypes',
                    'childContextTypes',
                    'mixins',
                    'statics',
                    'defaultProps',
                    'constructor',
                    'getDefaultProps',
                    'getInitialState',
                    'getChildContext',
                ],
                lifecycle: [
                    'componentWillMount',
                    'componentDidMount',
                    'componentWillReceiveProps',
                    'shouldComponentUpdate',
                    'componentWillUpdate',
                    'componentDidUpdate',
                    'componentWillUnmount',
                ],
                render: [
                    '/^render.+$/',
                    'render',
                ],
            },
        }],
    },
};
```

**What did you do? Please include the actual source code causing the issue.**

```js
// Redux reducer, pertinent bit:
let newSections = {}; // This gets an error from prefer-const

action.sections.forEach((sectionDefinition) => {
    const sectionID = sectionDefinition.id.toLowerCase() || 'general';
    // add to existing section or create new one
    const existingSection = state.sections[sectionID]
        || {
            title: sectionDefinition.title,
            settings: [],
        };

    // assignment !
    newSections[sectionID] = {
        ...existingSection,
        settings: Array.concat(existingSection.settings, sectionDefinition.settings),
    };
});

return {
    ...state,
    sections: {
        ...state.sections,
        ...newSections,
    },
};
```

**What did you expect to happen?**
newSections should not be an error of pre-const as its used after the function in the same scope where it was defined.

**What actually happened? Please include the actual, raw output from ESLint.**
SettingsReducer.js: 'newSections' is never reassigned, use 'const' instead. (prefer-const)

**Comment by gandazgul** [comment]
Apparently, using const for objects is not intuitive in that it allows adding keys to the object, so using const here will work and eslint is right in throwing the error.
